x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/TransportCreateApiKeyAction.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.action;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyAction;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.CreateApiKeyResponse;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.security.authc.TokenService;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+
+import java.util.ArrayList;
+
+/**
+ * Implementation of the action needed to create an API key
+ */
+public final class TransportCreateApiKeyAction extends HandledTransportAction<CreateApiKeyRequest, CreateApiKeyResponse> {
+
+    private final TokenService tokenService;
+    private final CompositeRolesStore compositeRolesStore;
+    private final SecurityContext securityContext;
+
+    @Inject
+    public TransportCreateApiKeyAction(Settings settings, TransportService transportService, ActionFilters actionFilters,
+                                       TokenService tokenService, CompositeRolesStore compositeRolesStore, SecurityContext context) {
+        super(settings, CreateApiKeyAction.NAME, transportService, actionFilters,
+            (Writeable.Reader<CreateApiKeyRequest>) CreateApiKeyRequest::new);
+        this.tokenService = tokenService;
+        this.compositeRolesStore = compositeRolesStore;
+        this.securityContext = context;
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateApiKeyRequest request, ActionListener<CreateApiKeyResponse> listener) {
+        final Authentication authentication = securityContext.getAuthentication();
+        if (authentication == null) {
+            listener.onFailure(new IllegalStateException("authentication is required"));
+        } else if (request.getRoleDescriptors() == null || request.getRoleDescriptors().isEmpty()) {
```

Comment:
Is it our intention that an api-key only ever stores a snapshot of the roles/privileges, and does not reflect changes to the user's access?
I can see how that's consistent, but we'll need to make it really clear.

Reply:
Yes that is the intention

Reply:
I think this is fine, but we definitely need to think about how we expect administrators to handle existing api-keys when a user's privileges are downgraded.
There's a fairly clear attack vector to grant yourself an API key with all your current privileges just in case your access is every revoked.
We dont need an answer now, but we do need to think it through.

Reply:
That's a very valid point. Could we check on Authentication that the creator's serialized roles are still a subset of current ones?